### DISCUSSION TITLE
How can I send MIDI message to a hardware device when a particular MIDI note is played as input?

### DISCUSSION THREAD

**Discussion by tjwrona** [helgoboss/helgobox#1206] [Q&A] [answered]
I am trying to set up a mapping using ReaLearn so when a particular MIDI note is played it will send a MIDI message to my Neural DSP Quad Cortex to switch the current preset I am using.

The goal I am trying to achieve here is to find a way to map MIDI messages sent from Guitar Pro to automatically switch tones on my quad cortex in realtime as the song plays. Guitar Pro is very limited in how it can interact with Reaper via MIDI, but I was thinking I would set up a track in guitar pro that just sends a specific MIDI note to Reaper and then have ReaLearn detect if that note was played and use it to send a MIDI message the the Quad Cortex to switch tones.

Here is a screenshot of my attempted mapping:
![image](https://github.com/user-attachments/assets/12120083-e314-4f08-bad5-80bf42bb7999)

If I try to change the "Value" parameter in the target I get the following error: Control: Failed to hit target: sending to device input is only possible in response to a MIDI source event coming from a MIDI device [invocation = 97, target = 1, control value = 1] (direct control)

**Comment by helgoboss** [comment]
You need to choose "Feedback output" instead of "Input" (and set ReaLearn's output accordingly).

Isn't it too much hassle though? Why not send the MIDI message to the device from Guitar Pro?

**Comment by tjwrona** [comment]
Thanks for the quick response! As far as I know, Guitar Pro is very limited in it's MIDI functionality and is not capable of sending specific MIDI CC commands. If I just make a track and map some notes on it, I can send those MIDI notes over to Reaper and (hopefully) use ReaLearn to translate those notes into the CC commands that I need to control my hardware device.

I tried configuring it as "feedback output" instead and I'm still having an issue where the "value" I'm trying to send keeps getting reset to 0 no matter what I set it to
![image](https://github.com/user-attachments/assets/0912f8d4-83db-48cb-ad08-c57311c11036)

**Comment by tjwrona** [comment]
I was able to get it working by setting the "Target Min/Max" settings both to the desired CC value in the "Glue" section! Still not sure why I can't just specify the value in the target section itself though. It just immediately resets to 0 no matter what I do.

**Comment by helgoboss** [comment]
Probably Guitar Pro sends a note-on followed by an immediate note-off. That would explain the immediate reset to 0. The correct way to handle that is to ignore note-off is to use "Press only" or to set source min to 1. Setting target min/max to a particular value will probably lead to duplicate MIDI messages.

You can actually set the concrete value in the target. In the "Pattern" field.

**Reply by tjwrona** [reply]
Even if I set source min to 1 and set "press only" it still won't let me set any value other than 0 in the "Value" field of the target. If I change it, the second I click "OK" it resets it back to 0
![image](https://github.com/user-attachments/assets/2fa58519-a785-4ba5-8909-c6bfacc0b6ea)

**Reply by helgoboss** [reply]
The "Value" field represents the **current** value of the target. So this is simply the wrong field to use.

You need to adjust the "Pattern" field. E.g. you would have to enter "B0 2B 0A", if you wanted to send CC value 10. Number 10 is 0A in the hexadecimal system. It's described in the manual.

But the alternative of setting target min and max to the desired CC value also works and is easier if you don't know the hexadecimal system.

In both cases, you should use "Press only" in order to avoid a message to be sent when a note-off message arrives.

**Reply by tjwrona** [reply]
Awesome thank you, I gave that a try and that works as expected.

**Comment by tjwrona** [comment, marked as answer]
Problem resolved, see above comments for the details :)
